### PR TITLE
Allow eager loading relationships on Aggregated models

### PR DESCRIPTION
**Pull request by JayBizzle**
This PR will allow you to define relationships that you would like to load on your Aggregated models.

Currently you can do something like this...
```php
$result = News::search($query)->get()->load('author');
```
...but this requires all your searchable models to have an `author` relationship.

We currently have an index that has `Company` and `Employee` models.

When getting search results, we want to eager load various (and different) relationships on each type of result.

This PR will allow you to define a relationship model map on the Aggregator class like this...

```php
<?php

namespace App\Search;

use Algolia\ScoutExtended\Searchable\Aggregator;

class News extends Aggregator
{
    /**
     * The names of the models that should be aggregated.
     *
     * @var string[]
     */
    protected $models = [
        \App\Models\Company::class,
        \App\Models\Employee::class,
    ];

    /**
     * Map of model relations to load.
     *
     * @var string[]
     */
    protected $relations = [
        \App\Models\Company::class => ['address'],
        \App\Models\Employee::class => ['photo', 'profile'],
    ];
}
```

Let me know what you think and if this has any legs, and if so, I can get some tests written for it.

Thanks

### CODE REVIEW

File: src/Searchable/ModelsResolver.php [hunk 1]
```diff
@@ -29,7 +29,7 @@ final class ModelsResolver
      * Get a set of models from the provided ids.
      *
      * @param \Laravel\Scout\Builder $builder
-     * @param  string|object $searchable
+     * @param  object $searchable
```
Comment (JayBizzle) [inline src/Searchable/ModelsResolver.php]: Pretty sure this is never a string?

File: src/Searchable/Aggregator.php [hunk 2]
```diff
@@ -41,6 +41,13 @@ abstract class Aggregator implements SearchableCountableContract
      */
     protected $model;
 
+    /**
+     * Map of model relations to load.
+     *
+     * @var mixed[]
+     */
+    protected $relations = [];
```
Comment (JayBizzle) [inline src/Searchable/Aggregator.php]: Relations can work like with an array of arrays or strings

```php
$relations = [
    \App\Models\Company::class => ['this', 'that'],
];
```

or...
```php
$relations = [
    \App\Models\Company::class => 'other',
];
```

File: src/Searchable/ModelsResolver.php [hunk 3]
```diff
@@ -54,7 +54,7 @@ public function from(Builder $builder, $searchable, array $ids): Collection
 
             if (in_array(Searchable::class, class_uses_recursive($model), true)) {
                 if (! empty($models = $model->getScoutModelsByIds($builder, $modelKeys))) {
-                    $instances = $instances->merge($models);
+                    $instances = $instances->merge($models->load($searchable->getRelations($modelClass)));
```
Comment (nunomaduro) [inline src/Searchable/ModelsResolver.php]: What happens if the relations are an empty array?

File: src/Searchable/Aggregator.php [hunk 4]
```diff
@@ -41,6 +41,13 @@ abstract class Aggregator implements SearchableCountableContract
      */
     protected $model;
 
+    /**
+     * Map of model relations to load.
```
Comment (nunomaduro) [inline src/Searchable/Aggregator.php]: What do you think about this description: `The relationships per model that should be loaded.`.

File: src/Searchable/ModelsResolver.php (hunk 3 above)
Comment (JayBizzle) [reply src/Searchable/ModelsResolver.php]: No errors and nothing will get loaded

File: src/Searchable/Aggregator.php (hunk 4 above)
Comment (JayBizzle) [reply src/Searchable/Aggregator.php]: Yes, i'd be happy with that 👍

**Comment by JayBizzle** [comment]
> 1. The developer will want always load their relations? What is the performance issue related to it?

The feature is entirely optional. Performance wise, i am not quite sure what you are asking as obviously eager-loading relations when possible is of great benefit to the performance.

> 2. If everything is good, do you mind of adding tests on this?

Yes, i will definitely do that once we have ironed out the details 👍

**Comment by nunomaduro** [comment]
Ok. You can advance for tests. Thanks for the hard work on this!

**Comment by JayBizzle** [comment]
Merged tests as requested 👍

**Pull request merged**